lightning/src/ln/channelmanager.rs

```diff
@@ -3515,32 +3534,43 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 				session_priv_bytes.copy_from_slice(&session_priv[..]);
 				let mut outbounds = self.pending_outbound_payments.lock().unwrap();
 				if let hash_map::Entry::Occupied(mut payment) = outbounds.entry(payment_id) {
-					let found_payment = !payment.get().is_fulfilled();
-					let fee_paid_msat = payment.get().get_pending_fee_msat();
-					payment.get_mut().mark_fulfilled();
+					let mut pending_events = self.pending_events.lock().unwrap();
+					if !payment.get().is_fulfilled() {
+						let payment_hash = PaymentHash(Sha256::hash(&payment_preimage.0).into_inner());
+						let fee_paid_msat = payment.get().get_pending_fee_msat();
+						pending_events.push(
+							events::Event::PaymentSent {
+								payment_id: Some(payment_id),
+								payment_preimage,
+								payment_hash,
+								fee_paid_msat,
+							}
+						);
+						payment.get_mut().mark_fulfilled();
+					}
+
 					if from_onchain {
 						// We currently immediately remove HTLCs which were fulfilled on-chain.
 						// This could potentially lead to removing a pending payment too early,
 						// with a reorg of one block causing us to re-add the fulfilled payment on
 						// restart.
 						// TODO: We should have a second monitor event that informs us of payments
 						// irrevocably fulfilled.
-						payment.get_mut().remove(&session_priv_bytes, Some(&path));
+						if payment.get_mut().remove(&session_priv_bytes, Some(&path)) {
+							let payment_hash = Some(PaymentHash(Sha256::hash(&payment_preimage.0).into_inner()));
+							pending_events.push(
+								events::Event::PaymentPathSuccessful {
+									payment_id,
+									payment_hash,
+									path,
+								}
```

Comment:
Hmmm, its definitely success, though. We could add a flag for it, but because routing is generally done via the `list_usable_channels` list we won't consider the now-closed channel anyway.

Reply:
Could this also result from another channel along the path?

From a scoring perspective, I'd imagine we'd want to score positively any channel that was used to route the HTLC to the payee. And any closed channels should be avoided through other means.

Reply:
I believe this can only result from one of our own channel closes. Could document that `path` may contain a closed channel in rare cases.

Reply:
Yea, this specific case, but the general case of "we got an accept because of an on-chain claim" is also an issue (and hits the other case). I don't think we can detect the difference there (aside from the HTLC fulfill being really slow), so I'm not sure there's anything we can do in the general case.